src/contract/deploy.rs

```diff
@@ -58,30 +58,34 @@ where
     pub fn new(web3: Web3<T>, context: I::Context) -> Self {
         let net = web3.net();
         DeployedFuture {
-            args: Some((web3.into(), context)),
+            args: Some((web3, context)),
             network_id: net.version().compat(),
             _instance: PhantomData,
         }
     }
 }
 
+impl<T, I> Unpin for DeployedFuture<T, I>
```

Comment:
Shouldn't this be autogenerated for us if CompatCallFuture<T, String> is Unpin? 

Is it because the T we are using are not Unpin? If that's the case, are we sure web3 doesn't use give out `Pin<T>` internally which could make it invalid for us to more the DeployedFuture Unpin?

Reply:
I think the idea here is that since we never pin `Web3<T>` it is not our responsibility to uphold any pinning guarantees. If `Web3<T>` would pin a `T: !Unpin` internally for some operation, then it would require to write `unsafe` code, and it would be responsible for upholding the pinning guarantee.

Reply:
And yes, `CompatCallFuture` wasn't auto-implementing `Unpin` because of `T` in `Web3<T>`.

Reply:
>... it would require to write unsafe code, and it would be responsible for upholding the pinning guarantee.

My understanding is that since they mark T !Unpin, they could e.g. be giving out a Pin<T> via unsafe code without risking any undefined behavior. However, if we then marked the composite struct Unpin, we could actually end up moving T which might have been given out by the lower layer and thus cause undefined behaviour.

But I'm not sure this logic makes sense, since we could already move the T from a plain mutable instance without any Pin/Unpin.

Reply:
Its not that `T: !Unpin` but more that `T: ?Unpin` in that it can be `Unpin` since there are no bounds on T (besides being `Transport + Clone`). And, `Web3` is defined something like this: `struct Web3<T: Transport> { transport: T }` so our futures can't auto implement `Unpin` because `Web3<T>` can't because `T` may be unpin (as there is no trait bound).

In practice, none of the transports are `!Unpin`, the following compiles:
```
fn main() {
    use web3::transports::*;
    fn assert_unpin<T: Unpin>() {}
    assert_unpin::<Http>();
    assert_unpin::<WebSocket>();
    assert_unpin::<Ipc>();
    assert_unpin::<Batch<Http>>();
}
```

That being said, suppose that there is an transport that is `T: !Unpin` (which is totally possible) our future always receives an owned `Web3<T>`, which is not pinned and as such, our instance of `Web3` is completely free to move around. If there were for example a method `Web3::do_something(self: Pin<&mut Self)`, then that would require the future to pin the `Web3` instance. As soon as the `Web3` instance is pinned (which would be an `unsafe` thing since `T: !Unpin` in our example), then it would be the future's responsibility once the `Web3` is pinned, to ensure that it never moves.

If a transport were self referential (for example it kept some shared buffers or something) then it most probably would need to be constructed in a way such that it is always pinned. And you would only ever get a `Pin<Box<T>>` to it. In that case, the `Pin<Box<T>>` would be safe to move around (as it would effectively just be copying a pointer and not the data being pointed to).

From everything I understood, `T: !Unpin` can move freely until pinned. This is why, in the `futures` crate you have something like this:
```
pub struct Ready<T>(Option<T>);
impl<T> Unpin for Ready<T> {}
```
(https://docs.rs/futures-util/0.3.1/src/futures_util/future/ready.rs.html#8-10)

Note that if we ever required a pinned `Web3<T>`, then the `pin_project` macro would automatically add `where Web3<T>: Unpin` as a constraint to the future's `Unpin` implementation. In that case, where `Web3<T>` requires pinning to be used, then indeed the future would only be `Unpin` if `Web3<T>: Unpin` which is equivalent to `T: Unpin`.

Reply:
This is another big pro to using `pin_project` as we don't have to remember to add these trait bounds in the case we need additional pinning on fields in our struct (most commonly this will happen if another future was added to the future combinator).

Also, since using `pin_project`, we no longer `impl Unpin` for anything (we defer that to the crate as it does it programmatically and makes sure that we don't forger to update anything). Check it out:
```
$ rg Unpin
$ rg Pin::new
$ 
```